migrations/versions/1060_add_external_id_column.py

```diff
@@ -0,0 +1,77 @@
+"""Add external ID column
+
+Revision ID: 1050
+Revises: 1040
+Create Date: 2017-11-23 16:46:00.569239
+
+"""
+from alembic import op
+import json
+import random
+import sqlalchemy as sa
+from sqlalchemy.sql import text
+
+# revision identifiers, used by Alembic.
+revision = '1060'
+down_revision = '1050'
+
+DIRECT_AWARD_AUDIT_TYPES = ('create_project', 'create_project_search', 'lock_project', 'downloaded_project')
+
+
+def random_positive_external_id():
+    return random.randint(2 ** 60, 2 ** 63 - 1)
+
+
+def upgrade():
+    op.add_column('direct_award_projects', sa.Column('external_id', sa.BIGINT()))
+    op.create_index(op.f('ix_direct_award_projects_external_id'), 'direct_award_projects', ['external_id'], unique=True)
+
+    # Manually generate random IDs for all existing saved searches
+    project_connection = op.get_bind()
+    project_id_internal_to_external = {}
+    for row_id, *_ in project_connection.execute('select id from direct_award_projects'):
+        # Some logic to ensure all IDs generated are unique; keep track so we can update AuditEvents.
+        random_id = random_positive_external_id()
+        while random_id in project_id_internal_to_external:
+            random_id = random_positive_external_id()
+        project_id_internal_to_external[row_id] = random_id
+
+        project_connection.execute(text('update direct_award_projects set external_id = :external_id where id = :id'),
+                                   {'external_id': random_id, 'id': row_id})
+
+    # Now that all rows have an external ID, generate the unique constraint for them.
+    op.create_unique_constraint('uq_direct_award_projects_external_id', 'direct_award_projects', ['external_id'])
+
+    # Inject external IDs into existing Audit Events to allow them to be searched by the reference that we are likely
+    # to be given e.g. by support enquiries.
+    audit_connection = op.get_bind()
+    for audit_id, audit_data, project_id in audit_connection.execute(
+            text('select id, data, object_id from audit_events where type IN :audit_types'),
+            {'audit_types': DIRECT_AWARD_AUDIT_TYPES}
+    ):
+        # If for some hellish reason we are re-running this after external IDs already exist, our audit events should
+        # keep the original external ID (ie not be overwritten).
+        new_audit_data = json.dumps({'projectExternalId': project_id_internal_to_external[project_id], **audit_data})
+        audit_connection.execute(text('update audit_events set data = (:new_audit_data)::json where id = :audit_id'),
+                                 {'new_audit_data': new_audit_data, 'audit_id': audit_id})
+
+
+def downgrade():
+    op.drop_index(op.f('ix_direct_award_projects_external_id'), table_name='direct_award_projects')
+    op.drop_column('direct_award_projects', 'external_id')
+
+    # I don't think we ought to delete the external ID from the Audit Event on a downgrade... But if we do, here's
```

Comment:
Why not?  The external IDs would be meaningless once they no longer exist in the projects table?

Reply:
In theory, if we ever did a downgrade migration on production, the IDs _would have_ existed and users would have seen them. They might start sending in support requests saying "Where's my project? It was reference `xyz` but that seems to not work anymore."

Maybe. (of course there are other ways they could find the search they're after, so maybe that's a bad example...)

But if we keep them, we would always be able to look up by the external ID.

Anyway I wasn't sure what we _should_ do, hence the comment and writing the SQL. If we should get rid of them and make it as if it never happened, I'll uncomment it.

Reply:
Be interested in your thoughts @idavidmcdonald

Reply:
It's an unlikely scenario, but I guess there's no harm in keeping them. And would save this scary looking update of all the audit JSON from existing.